java/simplesaga-action/src/main/java/io/simplesource/saga/action/async/AsyncApp.java

```diff
@@ -19,57 +9,56 @@
+ * AsyncApp is an Action Processor that executes arbitrary asynchronous functions as action requests.
+ *
+ * @param <A> common representation form for all action commands (typically Json / GenericRecord for Avro).
+ */
 public final class AsyncApp<A> {
 
     private final Logger logger = LoggerFactory.getLogger(AsyncApp.class);
-    private final List<AsyncTransformer<A>> transformers = new ArrayList<>();
     private final List<Supplier<Integer>> closeHandlers = new ArrayList<>();
 
     private final List<TopicCreation> expectedTopics;
     private final TopicConfig actionTopicConfig;
     private final ActionProcessorSpec<A> actionSpec;
     private ScheduledExecutorService executor;
-
-    @Value
-    private static class AsyncTransformerInput<A> {
-        final StreamsBuilder builder;
-        final KStream<UUID, ActionRequest<A>> actionRequests;
-        final KStream<UUID, ActionResponse> actionResponses;
-        final ScheduledExecutorService executor;
-    }
-
-    private interface AsyncTransformer<A> {
-        Function<Properties, AsyncTransform.AsyncPipe> apply(AsyncTransformerInput<A> input);
-    }
+    private AsyncTopologyBuilder<A> topologyBuilder;
+    private Set<ExecutorService> executorServices = new HashSet<>();
 
     public AsyncApp(ActionSerdes<A> actionSerdes, TopicConfigBuilder.BuildSteps topicBuildFn) {
         List<String> expectedTopicList = new ArrayList<>(TopicTypes.ActionTopic.all);
         expectedTopicList.add(TopicTypes.ActionTopic.requestUnprocessed);
 
         actionTopicConfig = TopicConfigBuilder.buildTopics(expectedTopicList, new HashMap<>(), new HashMap<>(), topicBuildFn);
-        actionSpec = new ActionProcessorSpec<A>(actionSerdes);
+        actionSpec = new ActionProcessorSpec<>(actionSerdes);
+        topologyBuilder = new AsyncTopologyBuilder<>(actionSpec, actionTopicConfig);
         expectedTopics = TopicCreation.allTopics(actionTopicConfig);
 
     }
 
     public <I, K, O, R> AsyncApp<A> addAsync(AsyncSpec<A, I, K, O, R> spec) {
-        AsyncTransformer<A> transformer = input -> {
-            AsyncContext<A, I, K, O, R> ctx = new AsyncContext<>(actionSpec, actionTopicConfig.namer, spec, input.executor);
-            // join the action request with corresponding prior command responses
-            AsyncStream.addSubTopology(ctx, input.actionRequests, input.actionResponses);
-
-            return AsyncTransform.async(ctx);
-        };
-        transformers.add(transformer);
         spec.outputSpec.ifPresent(oSpec -> expectedTopics.addAll(oSpec.topicCreations()));
+
+        topologyBuilder.onBuildTopology((topologyContext) -> {
+            ScheduledExecutorService usedExecutor = executor != null ? executor : Executors.newScheduledThreadPool(1);
+            executorServices.add(usedExecutor);
+            AsyncContext<A, I, K, O, R> async = new AsyncContext<>(actionSpec, actionTopicConfig.namer, spec, usedExecutor);
+            topologyBuilder.addSubTopology(topologyContext, async);
```

Comment:
Having an object with a with a method that takes a listener (`topologyBuilder.onBuildTopology` and then calling a method on that object (`topologyBuilder.addSubTopology`) in the listener implementation is a confusing pattern.

Reply:
I'd prefer if the `addSubTopology` method were a static method, or implementation was split out into a separate class. Then you could collapse the `AbstractActionTopologyBuilder` - the need for a class hierarchy would go away.

Reply:
Fixed